File: week-2/hw_5.py
def maxZeros(nums):
    '''
    Time Complexity:
        O(maxZeros(nums)): 
            O(n) + O(n)*O(1) + O(n) + O(n)*O(1) + O(n) => O(n) in average-case.

    reference: https://wiki.python.org/moin/TimeComplexity
    '''
    # str.join(): O(n) in average-case.
    # map(): O(n) in average-case.
    # int to str casting: Normally it should be O(n^2), 
    #                     which is O(n) for division and remainder operations to find n digits, 
    #                     and O(n) for each arithmetic operation.
    #                     However the digits we have here would be whether 1 or 0, 
    #                     The time complexity might be treated as O(1).
    # split(): O(n) in average-case.
    nums_lt = ''.join(map(str, nums)).split('1')

    # map(): O(n) in average-case.
    # len(): O(1) in average-case.
    nums_len = map(lambda elem: len(elem), nums_lt)

    # max(): O(n) in average-case.
    return max(nums_len)


print(maxZeros([0, 1, 0, 0]))
print(maxZeros([1, 0, 0, 0, 0, 1, 0, 1, 0, 0]))
print(maxZeros([1, 1, 1, 1, 1]))
print(maxZeros([0, 0, 0, 1, 1]))

File: week-2/hw_5.js
function maxZeros(nums) {
    /*
    Time Complexity:
        O(maxZeros(nums)): 
            O(n) + O(n) + O(n)*O(1) + O(n) => O(n) in average-case.

    reference: 
        1. https://tc39.es/ecma262/#sec-array.prototype.join
        2. https://tc39.es/ecma262/#sec-string.prototype.split
        3. https://tc39.es/ecma262/#sec-array.prototype.map
        4. https://tc39.es/ecma262/#sec-math.max
    */

    // Array.prototype.join(): O(n) in average-case.
    // String.prototype.split(): O(n) in average-case.
    let nums_arr = nums.join('').split('1');

    // Array.prototype.map(): O(n) in average-case.
    // Array.prototype.length: O(1) in average-case.
    let nums_len = nums_arr.map(elem => elem.length);

    // Math.max(): O(n) in average-case.
    return Math.max(...nums_len);
}


console.log(maxZeros([0, 1, 0, 0]));
console.log(maxZeros([1, 0, 0, 0, 0, 1, 0, 1, 0, 0]));
console.log(maxZeros([1, 1, 1, 1, 1]));
console.log(maxZeros([0, 0, 0, 1, 1]));